File: java/src/patterns/creational/singleton/StaticBlockSingleton.java
package patterns.creational.singleton;

class StaticBlockSingleton {
    // single static instance
    private static StaticBlockSingleton instance;

    // private constructor prevents object instantiation
    private StaticBlockSingleton() {
        // init logic here
    }

    // executed (only once) when the class is loaded
    static {
        try {
            instance = new StaticBlockSingleton();
        } catch (Exception e) {
            throw new RuntimeException("Error creating instance", e);
        }
    }

    public static StaticBlockSingleton getInstance() {
        return instance;
    }
}

File: python/src/patterns/creational/singleton/static_block_singleton.py
class StaticBlockSingleton:
    _instance = None

    def __init__(self) -> None:
        # prevent multiple instantiation by raising an exception
        if StaticBlockSingleton._instance is not None:
            raise Exception("Cannot instantiate directly.")
        # init logic here

    @staticmethod
    def get_instance() -> "StaticBlockSingleton":
        return StaticBlockSingleton._instance

# instance created immediately on module import
try:
    StaticBlockSingleton._instance = StaticBlockSingleton()
except Exception as e:
    raise Exception("Error creating instance", e)
